### PR TITLE
Feature Addition: Improved status feedback during profile move

### PR DESCRIPTION
**Pull request by triosl**
These changes were developed in order to improve the status feedback during a profile move to be consumed by malcolm.  However, I feel these changes are probably generally useful and so should be merged into the main branch.

Here is a summary of the changes:
1. Add Automation1MotorController::poll override to monitor the currently executing profile
2. Post error messages to message PVs when in profileBuild/Execute/Readback
3. Fix different units for readback and position arrays.

     This probably doesn't belong in this PR, so I can make another with this change if required, but it seems to me that the 
     MovePt Aeroscript command is expecting positions in counts and not user units, whereas, the data collection returns
     positions in user units.  

4. Split Automation1_Task_ProgramRun into Automation1_Task_ProgramLoad and Automation1_Task_ProgramStart.  Automation1_Task_ProgramLoad is now run during profileBuild and not profileExecute.

    I am not sure if this is correct, but it seems to me that compiling the program should be part of the build step rather than execute step.  I can remove this change if there is some reason that the compilation needs to happen during the execute step.

### CODE REVIEW

**Comment by rakaerotech** [comment]
Does this PR obsolete Issue #3?

**Comment by thomasives** [comment]
Yes I believe it does solve #3

**Comment by rakaerotech** [comment]
> Fix different units for readback and position arrays ... it seems to me that the
MovePt Aeroscript command is expecting positions in counts and not user units

This should not be the case. **MovePt** expects input to be in user units (either the default of primary user units or optionally secondary user units depending on **SetupTaskDistanceUnits**). That being said I don't fully grok the EPICS units system, and I don't know the units of **profilePositions_**, so your change on line 363 might be correct. All I know is that **MovePt** does not accept hardware counts, it expects user units (e.g. mm).

> Split Automation1_Task_ProgramRun into Automation1_Task_ProgramLoad and Automation1_Task_ProgramStart. Automation1_Task_ProgramLoad is now run during profileBuild and not profileExecute ... I am not sure if this is correct, but it seems to me that compiling the program should be part of the build step rather than execute step.

This is arguably a more correct thing to do and I can get behind the change, however, the **Automation1_Task_ProgramStart** in **profileExecute** doesn't know what program its about to start, it could be starting any AeroScript program loaded onto task 2 (**PROFILE_MOVE_TASK_INDEX**). This increases the likelihood that EPICS starts a program it wasn't expecting. If EPICS is the only thing interacting with Automation1 then this is mostly moot but worth thinking about. The best solution would be to only compile the program **profileBuild** then load and start in **profileExecute**, but we currently don't have a way to only compile in the C API.

Long story short: I'm ok with this as long as we have a degree of confidence that other processes aren't loading AeroScript programs onto Automation1 tasks.

File: automation1App/src/Automation1MotorController.cpp [hunk 1]
```diff
@@ -606,34 +588,100 @@ asynStatus Automation1MotorController::readbackProfile()
...
+        if (!Automation1_DataCollection_GetStatus(controller_, &dataCollectionStatus))
+        {
+            logApiError(profileExecuteMessage_, "Failed to get data collection status");
+            pollOk = false;
+            goto done;
+        }
+
+        setIntegerParam(profileCurrentPoint_, dataCollectionStatus.NumberOfRetrievedPoints);
+
+        if (!dataCollectionStatus.IsCollecting) {
```
Comment (rakaerotech) [inline automation1App/src/Automation1MotorController.cpp:624]: Instead of checking if data collection is still collecting, it might be more correct to check the task state of **PROFILE_MOVE_TASK_INDEX** via `Automation1TaskStatus.TaskState == Automation1TaskState_ProgramRunning`.

Or perhaps checking both task state *and* data collection status.

**Comment by triosl** [comment]
>This should not be the case. MovePt expects input to be in user units (either the default of primary user units or optionally secondary user units depending on SetupTaskDistanceUnits). That being said I don't fully grok the EPICS units system, and I don't know the units of profilePositions_, so your change on line 363 might be correct. All I know is that MovePt does not accept hardware counts, it expects user units (e.g. mm)

In this case, I will remove this change from the PR as I also don't really understand things from either side and it doesn't belong here anyway.  The change seems to have made the units consistent between the readback arrays and the positions arrays for the single test case I was using, however, I don't think this is enough to go on.  I will try to concretely understand why this change was working for me and create another PR for this when I do.

>This is arguably a more correct thing to do and I can get behind the change, however, the Automation1_Task_ProgramStart in profileExecute doesn't know what program its about to start, it could be starting any AeroScript program loaded onto task 2 (PROFILE_MOVE_TASK_INDEX). This increases the likelihood that EPICS starts a program it wasn't expecting. If EPICS is the only thing interacting with Automation1 then this is mostly moot but worth thinking about. The best solution would be to only compile the program profileBuild then load and start in profileExecute, but we currently don't have a way to only compile in the C API.

I see, I think I hadn't appreciated that `Automation1_Task_ProgramLoad` meant "load the program" (:facepalm:) and not just compile.  

Does the controller use last modified times of files or something similar to determine when it needs to compile the file, or does it always compile when you load the program?  i.e. if we issued a `Automation_Task_ProgramLoad` in `profileBuild` and a `Automation1_Task_ProgramRun` in `profileExecute` would that avoid compiling for the execute step (given we don't change the file)?

If not then I think we should just leave it as it is as I doubt it matters too much and there are some situations where my version will break things.

**Comment by triosl** [comment]
I have removed the changes corresponding to points 3 and 4 from my original comment.  I will create a separate PR for point 3 when I have a better understanding of the unit systems.

I have also changed the check for when the profile move has finished to `Automation1TaskStatus.TaskState != Automation1TaskState_ProgramRunning` as suggested by @rakaerotech.  I don't think we need to check the collection status here, as we wait for the collection to finish at the start of `Automation1MotorController::readbackProfile`.

**Comment by rakaerotech** [comment]
These changes make sense to me from Automation1's perspective but, as usual, I can't really sign-off on the EPICS changes.

**Comment by kmpeters** [comment]
> I will create a separate PR for point 3 when I have a better understanding of the unit systems.
>
>> 3. Fix different units for readback and position arrays.
>> 
>> This probably doesn't belong in this PR, so I can make another with this change if required, but it seems to me that the
>> MovePt Aeroscript command is expecting positions in counts and not user units, whereas, the data collection returns
>> positions in user units.

The motor record allows users to work in a user coordinate system that is offset from the dial coordinate system.  The zero of the dial coordinate system is the same as the controller's zero position, even if the units are not the same.

The motor record sends (dial) positions, velocities and accelerations to the driver in units of steps for point-to-point moves.  The MRES is used to convert between EPICS dial units and steps.  The MRES can be changed by the user as long as the step size remains the same.  An example of why someone might do this is to work in inches instead of mm.

It is the driver's responsiblity to convert steps into the controller units.

The profile move arrays, however, are specified in EPICS user units.  The driver needs to convert the position array to steps using the EPICS MRES & OFF, and then convert from steps to controller units.  The ``asynMotorAxis::readbackProfile`` method handles the conversion from steps to EPICS user units for you:

https://github.com/epics-modules/motor/blob/8fdd35dc11f012182cea2d43340e7e53f7b7fbed/motorApp/MotorSrc/asynMotorAxis.cpp#L432-L437

**Comment by kmpeters** [comment]
I don't have any objections.  Is this pull request ready to be merged?
